securedrop_client/gui/widgets.py

```diff
@@ -789,52 +789,57 @@ def update(self, sources: List[Source]) -> List[str]:
         """
         Update the list with the passed in list of sources.
         """
-        # Delete widgets that no longer exist in source list
-        source_uuids = [source.uuid for source in sources]
+        sources_to_update = []
+        sources_to_add = []
+        for source in sources:
+            try:
+                if source.uuid in self.source_items:
+                    sources_to_update.append(source.uuid)
+                else:
+                    sources_to_add.append(source)
+            except sqlalchemy.exc.InvalidRequestError as e:
+                logger.debug(e)
+                continue
+
+        # Delete widgets for sources not in the supplied sourcelist
         deleted_uuids = []
-        for i in range(self.count()):
-            list_item = self.item(i)
-            list_widget = self.itemWidget(list_item)
+        sources_to_delete = [self.source_items[uuid] for uuid in self.source_items
+                             if uuid not in sources_to_update]
+        for source_item in sources_to_delete:
+            if source_item.isSelected():
+                self.setCurrentItem(None)
 
-            if list_widget and list_widget.source_uuid not in source_uuids:
-                if list_item.isSelected():
-                    self.setCurrentItem(None)
+            source_widget = self.itemWidget(source_item)
+            self.takeItem(self.row(source_item))
+            source_widget.deleteLater()
+            if source_widget.source_uuid in self.source_items:
```

Comment:
general question on `deleteLater()`: we're accessing an attribute on the `source_widget` object we just called `deleteLater()` on - do you know when that `deletion` event will get processed? I was wondering if it's possible for the deletion event to occur prior to this attribute access, my expectation is that the deletion event can only happen after this method returns (i.e. when control returns to the GUI event loop) but want to confirm to ensure this is safe.

Reply:
good find. you're right that we want to delete _after_ this